sshd-core/src/main/java/org/apache/sshd/common/channel/ChannelAsyncOutputStream.java

```diff
@@ -282,4 +435,43 @@ public boolean isSendChunkIfRemoteWindowIsSmallerThanPacketSize() {
     public void setSendChunkIfRemoteWindowIsSmallerThanPacketSize(boolean sendChunkIfRemoteWindowIsSmallerThanPacketSize) {
         this.sendChunkIfRemoteWindowIsSmallerThanPacketSize = sendChunkIfRemoteWindowIsSmallerThanPacketSize;
     }
+
+    // Marker type to avoid repeated buffering
+    private static class BufferedFuture extends IoWriteFutureImpl {
+
+        BufferedFuture(Object id, Buffer buffer) {
+            super(id, buffer);
+        }
+    }
+
+    // Collects state variables; access is always synchronized on the single instance per stream.
+    private static class WriteState {
```

Comment:
Let's make this *protected* (as well as its constructor, fields and methods) in the spirit of allowing users to overload anything they want in our code.

Reply:
Done.